src/Microsoft.ApiDesignGuidelines.Analyzers/Core/NonConstantFieldsShouldNotBeVisible.cs

```diff
@@ -32,8 +33,23 @@ public abstract class NonConstantFieldsShouldNotBeVisibleAnalyzer : DiagnosticAn
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
-        { 
-            
+        {
+            analysisContext.RegisterSymbolAction(obj =>
+            {
+                var field = (IFieldSymbol)obj.Symbol;
+
+                if (!field.IsPublic())
+                {
+                    return;
+                }
+
+                if (field.IsConst || (field.IsStatic && field.IsReadOnly))
+                {
+                    return;
+                }
+
+                obj.ReportDiagnostic(field.CreateDiagnostic(Rule));
```

Comment:
The first line of the description states: Static fields that are neither constants nor read-only are not thread-safe. Looks like it is specific to static fields. I can change my implementation to fit that.